File: Suspension matlab_org/Suspension matlab raw/four_wheel_load/VehicleDynamicsSimulator.m
classdef VehicleDynamicsSimulator
    properties
        g         % 重力加速度
        m         % 車重 (kg)
        h         % 重心高度 (m)
        W         % 車重力 (N)
        L_rear    % 後軸到重心距離 (m)
        L_front   % 前軸到重心距離 (m)
        L         % 軸距 (m)
        t_front   % 前輪距 (m)
        t_rear    % 後輪距 (m)
        K         % Ride rate 矩陣 (4x4)
        K_modal   % 模態剛性矩陣 (3x3)
        K_modal_coupled % 耦合模態剛性矩陣
        K_tire    % 輪胎剛性 (1x4)
    end
    
    methods
        % =====================================================================
        % Constructor (建構子) - 初始化參數
        % =====================================================================
        function obj = VehicleDynamicsSimulator(g_val)
            if nargin < 1
                obj.g = 9.81;
            else
                obj.g = g_val;
            end
            
            % 車輛基本參數定義
            obj.m = 360;
            obj.h = 0.286;
            obj.W = obj.m * obj.g;
            obj.L_rear = 0.744;
            obj.L_front = 0.806;
            obj.L = obj.L_front + obj.L_rear;
            
            obj.t_front = 1.25;
            obj.t_rear = 1.25;
            
            obj.K = [30000, 0, 0, 0;
                     0, 30000, 0, 0;
                     0, 0, 27000, 0;
                     0, 0, 0, 27000];
                 
            obj.K_modal = [3050, 0, 0;
                           0, 550, 0;
                           0, 0, 2500];
                       
            obj.K_modal_coupled = [3050, 0, 0, 0;
                                   0, 600 + 2900, -2900, 0;
                                   0, -2900, 500 + 2900, 0;
                                   0, 0, 0, 2500];
                               
            obj.K_tire = ones(1, 4) * 56000;
        end
        
        % =====================================================================
        % 內部輔助工具 (幾何、外力矩、Debug)
        % =====================================================================
        function [x_pos, y_pos, z_pos] = get_geometry(obj)
            x_front = obj.L_front;
            x_rear  = -obj.L_rear;
            
            y_left  = -obj.t_front / 2;
            y_right =  obj.t_front / 2;
            
            y_left_r  = -obj.t_rear / 2;
            y_right_r =  obj.t_rear / 2;
            
            x_pos = [x_front, x_front, x_rear, x_rear];
            y_pos = [y_left, y_right, y_left_r, y_right_r];
            z_pos = ones(1, 4);
        end
        
        function [Fz_val, Mx, My] = external_effects(obj, F_add, CF_rela)
            if isempty(F_add)
                Fz_val = 0; Mx = 0; My = 0;
                return;
            end
            Fx = F_add(1); Fy = F_add(2); Fz = F_add(3);
            if isempty(CF_rela)
                r = [0, 0, obj.h];
            else
                r = CF_rela;
            end
            Mx = r(2)*Fz - r(3)*Fy;
            My = r(3)*Fx - r(1)*Fz;
            Fz_val = Fz;
        end
        
        function [A, b, Mx_add, My_add] = get_Ab(obj, ax, ay, F_add, CF_rela)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            
            [x_pos, y_pos, ~] = obj.get_geometry();
            [Fz_add, Mx_add, My_add] = obj.external_effects(F_add, CF_rela);
            
            A = [1,        1,        1,        1;
                 y_pos(1), y_pos(2), y_pos(3), y_pos(4);
                 x_pos(1), x_pos(2), x_pos(3), x_pos(4)];
             
            b = [Fz_add - obj.m * obj.g;
                 obj.m * ay * obj.h + Mx_add;
                 obj.m * ax * obj.h + My_add];
        end
        
        function debug_balance(obj, ax, ay, F_add, N, x_pos, y_pos, Mx_add, My_add)
            if isempty(F_add), F_add = [0, 0, 0]; end
            az = -obj.g;
            
            fprintf('輪胎載重 (N):\n');
            fprintf('FL = %.1f, FR = %.1f, RL = %.1f, RR = %.1f\n', N(1), N(2), N(3), N(4));
            
            Fz = sum(N) + obj.m * az + F_add(3);
            fprintf('平衡確認 : Fz = %.6f N\n', Fz);
            
            % 確保這裡 N 是直向量 (4x1) 以利矩陣乘法
            Mx = y_pos * N + obj.m * ay * obj.h + Mx_add;
            My = x_pos * N + obj.m * ax * obj.h + My_add;
            
            fprintf('Roll moment balance (Mx) = %.6f Nm\n', Mx);
            fprintf('Pitch moment balance (My) = %.6f Nm\n', My);
        end
        
        % =====================================================================
        % 核心求解器 (Solvers)
        % =====================================================================
        function N = solve_cg(obj, ax, ay, F_add, CF_rela, check)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            if nargin < 6, check = false; end
            
            F_z_front_static = obj.W * (obj.L_rear / obj.L);
            F_z_rear_static  = obj.W * (obj.L_front / obj.L);
            
            dF_z_long = obj.m * obj.h * ax / obj.L;
            [~, Mx_add, My_add] = obj.external_effects(F_add, CF_rela);
            dF_z_long = dF_z_long + My_add / obj.L;
            
            dF_z_lat_f = (obj.m * ay * obj.h * (obj.L_rear / obj.L)) / obj.t_front;
            dF_z_lat_r = (obj.m * ay * obj.h * (obj.L_front / obj.L)) / obj.t_rear;
            
            dF_z_lat_f = dF_z_lat_f + (Mx_add * obj.L_rear / obj.L) / obj.t_front;
            dF_z_lat_r = dF_z_lat_r + (Mx_add * obj.L_front / obj.L) / obj.t_rear;
            
            N_fl = 0.5 * F_z_front_static - 0.5 * dF_z_long + dF_z_lat_f;
            N_fr = 0.5 * F_z_front_static - 0.5 * dF_z_long - dF_z_lat_f;
            N_rl = 0.5 * F_z_rear_static  + 0.5 * dF_z_long + dF_z_lat_r;
            N_rr = 0.5 * F_z_rear_static  + 0.5 * dF_z_long - dF_z_lat_r;
            
            N = [N_fl; N_fr; N_rl; N_rr]; % 回傳 4x1 直向量
            
            if check
                fprintf('\n=== CG Solver Results ===\n');
                [x_pos, y_pos, ~] = obj.get_geometry();
                obj.debug_balance(ax, ay, F_add, N, x_pos, y_pos, Mx_add, My_add);
            end
        end
        
        function N = solve_lsm(obj, ax, ay, F_add, CF_rela, check)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            if nargin < 6, check = false; end
            
            [A, b, Mx_add, My_add] = obj.get_Ab(ax, ay, F_add, CF_rela);
            % MATLAB 的 \ (backslash) 是最有效率且穩定的矩陣求解法
            % Python 的 A.T @ np.linalg.solve(A @ A.T, -b) 寫成 MATLAB 如下：
            N = A' * ((A * A') \ (-b));
            
            if check
                fprintf('\n=== LSM Solver Results ===\n');
                [x_pos, y_pos, ~] = obj.get_geometry();
                obj.debug_balance(ax, ay, F_add, N, x_pos, y_pos, Mx_add, My_add);
            end
        end
        
        function N = solve_lagrange(obj, ax, ay, F_add, CF_rela, check)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            if nargin < 6, check = false; end
            
            [x_pos, y_pos, ~] = obj.get_geometry();
            [A, b, Mx_add, My_add] = obj.get_Ab(ax, ay, F_add, CF_rela);
            
            N = obj.K * A' * ((A * obj.K * A') \ (-b));
            
            if check
                fprintf('\n=== Weighted Lagrange ===\n');
                obj.debug_balance(ax, ay, F_add, N, x_pos, y_pos, Mx_add, My_add);
            end
        end
        
        function N = solve_suspension(obj, ax, ay, F_add, CF_rela, check)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            if nargin < 6, check = false; end
            
            [x_pos, y_pos, ~] = obj.get_geometry();
            
            B = [ones(1, 4); y_pos; -x_pos]; 
            b = obj.m * [-obj.g; ay * obj.h; ax * obj.h];
            [Fz_add, Mx_add, My_add] = obj.external_effects(F_add, CF_rela);
            b = b + [Fz_add; Mx_add; My_add];
            
            A = [ones(1, 4); y_pos; x_pos];
            M = A * obj.K * B';
            q = M \ (-b);
            
            dz = B' * q;
            N = obj.K * dz;
            
            if check
                fprintf('\n=== Suspension Model ===\n');
                f_add_debug = F_add; if isempty(f_add_debug), f_add_debug = zeros(1,3); end
                obj.debug_balance(ax, ay, f_add_debug, N, x_pos, y_pos, Mx_add, My_add);
            end
        end
        
        function N = solve_decoupled(obj, ax, ay, F_add, CF_rela, check)
            if nargin < 4, F_add = []; end
            if nargin < 5, CF_rela = []; end
            if nargin < 6, check = false; end
            
            [x_pos, y_pos, ~] = obj.get_geometry();
            
            B = [ones(1, 4); y_pos; -x_pos];
            b = obj.m * [-obj.g; ay * obj.h; ax * obj.h];
            [Fz_add, Mx_add, My_add] = obj.external_effects(F_add, CF_rela);
            b = b + [Fz_add; Mx_add; My_add];
            
            A = [ones(1, 4); y_pos; x_pos];
            M = A * B' * obj.K_modal;
            q = M \ (-b);
            
            N = B' * (obj.K_modal * q);
            
            if check
                fprintf('\n=== Decoupled Suspension Model ===\n');
                f_add_debug = F_add; if isempty(f_add_debug), f_add_debug = zeros(1,3); end
                obj.debug_balance(ax, ay, f_add_debug, N, x_pos, y_pos, Mx_add, My_add);
            end
        end
    end
end

File: Suspension matlab/four_wheel_load/VehicleDynamicsSimulator/VehicleDynamicsSimulator.py
from __future__ import annotations

from pathlib import Path
import sys

import numpy as np

ROOT = next(p for p in Path(__file__).resolve().parents if (p / "suspension_utils.py").exists())
sys.path.insert(0, str(ROOT))

from suspension_utils import load_data


def _cfg(config: dict, name: str, default):
    return config.get(name, default)


class VehicleDynamicsSimulator:
    def __init__(self, g_val: float | None = None):
        data_file = Path(__file__).with_name("data.csv")
        config = load_data(data_file) if data_file.exists() else {}

        self.g = _cfg(config, "sim_g", 9.81) if g_val is None else g_val
        self.m = _cfg(config, "sim_m", 360)
        self.h = _cfg(config, "sim_h", 0.286)
        self.W = self.m * self.g
        self.L_rear = _cfg(config, "sim_L_rear", 0.744)
        self.L_front = _cfg(config, "sim_L_front", 0.806)
        self.L = self.L_front + self.L_rear
        self.t_front = _cfg(config, "sim_t_front", 1.25)
        self.t_rear = _cfg(config, "sim_t_rear", 1.25)

        self.K = np.diag(
            [
                _cfg(config, "sim_K_front_left", 30000),
                _cfg(config, "sim_K_front_right", 30000),
                _cfg(config, "sim_K_rear_left", 27000),
                _cfg(config, "sim_K_rear_right", 27000),
            ]
        )
        self.K_modal = np.diag(
            [
                _cfg(config, "sim_K_modal_heave", 3050),
                _cfg(config, "sim_K_modal_roll", 550),
                _cfg(config, "sim_K_modal_pitch", 2500),
            ]
        )
        coupled = _cfg(config, "sim_K_modal_coupled_roll_coupling", 2900)
        self.K_modal_coupled = np.array(
            [
                [_cfg(config, "sim_K_modal_coupled_heave", 3050), 0, 0, 0],
                [0, _cfg(config, "sim_K_modal_coupled_roll_base", 600) + coupled, -coupled, 0],
                [0, -coupled, _cfg(config, "sim_K_modal_coupled_pitch_base", 500) + coupled, 0],
                [0, 0, 0, _cfg(config, "sim_K_modal_coupled_pitch", 2500)],
            ],
            dtype=float,
        )
        self.K_tire = np.ones(4) * _cfg(config, "sim_K_tire", 56000)

    def get_geometry(self):
        x_front = self.L_front
        x_rear = -self.L_rear
        y_left = -self.t_front / 2
        y_right = self.t_front / 2
        y_left_r = -self.t_rear / 2
        y_right_r = self.t_rear / 2
        x_pos = np.array([x_front, x_front, x_rear, x_rear], dtype=float)
        y_pos = np.array([y_left, y_right, y_left_r, y_right_r], dtype=float)
        z_pos = np.ones(4)
        return x_pos, y_pos, z_pos

    def external_effects(self, F_add=None, CF_rela=None):
        if F_add is None or np.asarray(F_add).size == 0:
            return 0.0, 0.0, 0.0
        force = np.asarray(F_add, dtype=float).ravel()
        fx, fy, fz = force[:3]
        r = np.array([0.0, 0.0, self.h]) if CF_rela is None or np.asarray(CF_rela).size == 0 else np.asarray(CF_rela, dtype=float).ravel()
        mx = r[1] * fz - r[2] * fy
        my = r[2] * fx - r[0] * fz
        return fz, mx, my

    def get_Ab(self, ax, ay, F_add=None, CF_rela=None):
        x_pos, y_pos, _ = self.get_geometry()
        fz_add, mx_add, my_add = self.external_effects(F_add, CF_rela)
        A = np.array(
            [
                [1, 1, 1, 1],
                [y_pos[0], y_pos[1], y_pos[2], y_pos[3]],
                [x_pos[0], x_pos[1], x_pos[2], x_pos[3]],
            ],
            dtype=float,
        )
        b = np.array([fz_add - self.m * self.g, self.m * ay * self.h + mx_add, self.m * ax * self.h + my_add], dtype=float)
        return A, b, mx_add, my_add

    def debug_balance(self, ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add):
        force = np.zeros(3) if F_add is None or np.asarray(F_add).size == 0 else np.asarray(F_add, dtype=float).ravel()
        az = -self.g
        print("Wheel loads (N):")
        print(f"FL = {N[0]:.1f}, FR = {N[1]:.1f}, RL = {N[2]:.1f}, RR = {N[3]:.1f}")
        fz = np.sum(N) + self.m * az + force[2]
        print(f"Balance check : Fz = {fz:.6f} N")
        mx = y_pos @ N + self.m * ay * self.h + mx_add
        my = x_pos @ N + self.m * ax * self.h + my_add
        print(f"Roll moment balance (Mx) = {mx:.6f} Nm")
        print(f"Pitch moment balance (My) = {my:.6f} Nm")

    def solve_cg(self, ax, ay, F_add=None, CF_rela=None, check=False):
        f_z_front_static = self.W * (self.L_rear / self.L)
        f_z_rear_static = self.W * (self.L_front / self.L)
        _, mx_add, my_add = self.external_effects(F_add, CF_rela)

        df_z_long = self.m * self.h * ax / self.L + my_add / self.L
        df_z_lat_f = (self.m * ay * self.h * (self.L_rear / self.L)) / self.t_front + (mx_add * self.L_rear / self.L) / self.t_front
        df_z_lat_r = (self.m * ay * self.h * (self.L_front / self.L)) / self.t_rear + (mx_add * self.L_front / self.L) / self.t_rear

        N = np.array(
            [
                0.5 * f_z_front_static - 0.5 * df_z_long + df_z_lat_f,
                0.5 * f_z_front_static - 0.5 * df_z_long - df_z_lat_f,
                0.5 * f_z_rear_static + 0.5 * df_z_long + df_z_lat_r,
                0.5 * f_z_rear_static + 0.5 * df_z_long - df_z_lat_r,
            ]
        )
        if check:
            print("\n=== CG Solver Results ===")
            x_pos, y_pos, _ = self.get_geometry()
            self.debug_balance(ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add)
        return N

    def solve_lsm(self, ax, ay, F_add=None, CF_rela=None, check=False):
        A, b, mx_add, my_add = self.get_Ab(ax, ay, F_add, CF_rela)
        N = A.T @ np.linalg.solve(A @ A.T, -b)
        if check:
            print("\n=== LSM Solver Results ===")
            x_pos, y_pos, _ = self.get_geometry()
            self.debug_balance(ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add)
        return N

    def solve_lagrange(self, ax, ay, F_add=None, CF_rela=None, check=False):
        x_pos, y_pos, _ = self.get_geometry()
        A, b, mx_add, my_add = self.get_Ab(ax, ay, F_add, CF_rela)
        N = self.K @ A.T @ np.linalg.solve(A @ self.K @ A.T, -b)
        if check:
            print("\n=== Weighted Lagrange ===")
            self.debug_balance(ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add)
        return N

    def solve_suspension(self, ax, ay, F_add=None, CF_rela=None, check=False):
        x_pos, y_pos, _ = self.get_geometry()
        B = np.vstack([np.ones(4), y_pos, -x_pos])
        b = self.m * np.array([-self.g, ay * self.h, ax * self.h])
        fz_add, mx_add, my_add = self.external_effects(F_add, CF_rela)
        b = b + np.array([fz_add, mx_add, my_add])
        A = np.vstack([np.ones(4), y_pos, x_pos])
        q = np.linalg.solve(A @ self.K @ B.T, -b)
        N = self.K @ B.T @ q
        if check:
            print("\n=== Suspension Model ===")
            self.debug_balance(ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add)
        return N

    def solve_decoupled(self, ax, ay, F_add=None, CF_rela=None, check=False):
        x_pos, y_pos, _ = self.get_geometry()
        B = np.vstack([np.ones(4), y_pos, -x_pos])
        b = self.m * np.array([-self.g, ay * self.h, ax * self.h])
        fz_add, mx_add, my_add = self.external_effects(F_add, CF_rela)
        b = b + np.array([fz_add, mx_add, my_add])
        A = np.vstack([np.ones(4), y_pos, x_pos])
        q = np.linalg.solve(A @ B.T @ self.K_modal, -b)
        N = B.T @ (self.K_modal @ q)
        if check:
            print("\n=== Decoupled Suspension Model ===")
            self.debug_balance(ax, ay, F_add, N, x_pos, y_pos, mx_add, my_add)
        return N
